### PR TITLE
ICU-22332 bidiSkeleton and LTR- and RTL-confusabilities

### PR DESCRIPTION
**Pull request by eggrobin**
##### Checklist

- [x] Required: Issue filed: https://unicode-org.atlassian.net/browse/ICU-22332
- [x] Required: The PR title must be prefixed with a JIRA Issue number. 
- [x] Required: The PR description must include the link to the Jira Issue, for example by completing the URL in the first checklist item
- [x] Required: Each commit message must be prefixed with a JIRA Issue number. 
- [x] Issue accepted (done by Technical Committee after discussion)
- [x] Tests included, if applicable
- [x] API docs and/or User Guide docs changed or added, if applicable

### CODE REVIEW

**Comment by eggrobin** [comment]
(Note to self: I need to fix https://github.com/unicode-org/icu/blob/main/icu4c/source/test/depstest/dependencies.txt.)

**Comment by FrankYFTang** [comment]
I think you can fix your problem by the following patch to your PR
```
diff --git a/icu4c/source/test/depstest/dependencies.txt b/icu4c/source/test/depstest/dependencies.txt
index ca5027ed5c..86d023e203 100644
--- a/icu4c/source/test/depstest/dependencies.txt
+++ b/icu4c/source/test/depstest/dependencies.txt
@@ -920,7 +920,7 @@ group: charset_detector
 group: spoof_detection
     uspoof.o uspoof_build.o uspoof_conf.o uspoof_impl.o scriptset.o
   deps
-    uniset_props regex unorm uscript
+    uniset_props regex unorm uscript ubidi
 
 group: alphabetic_index
     alphaindex.o
```

File: icu4c/source/i18n/unicode/uspoof.h [hunk 1]
```diff
@@ -1159,6 +1179,56 @@ uspoof_areConfusable(const USpoofChecker *sc,
                      const UChar *id2, int32_t length2,
                      UErrorCode *status);
 
+/**
```
Comment (FrankYFTang) [inline icu4c/source/i18n/unicode/uspoof.h:1182]: you need to wrap these new functions with

```
#ifndef U_HIDE_DRAFT_API
```
and
```
#endif /* U_HIDE_DRAFT_API */
```
See https://github.com/unicode-org/icu/commit/27181e36a6bed372482e5a11c25009fa4dd971f1#diff-b9219dc4c2592180776f39139e72abfd4a9dc27cb0a7adb128b9c76b7bd7d65eR16 as example

File: icu4c/source/i18n/unicode/uspoof.h [hunk 2]
```diff
@@ -1192,14 +1262,44 @@ uspoof_areConfusableUTF8(const USpoofChecker *sc,
                          const char *id2, int32_t length2,
                          UErrorCode *status);
 
-
+/**
```
Comment (FrankYFTang) [inline icu4c/source/i18n/unicode/uspoof.h:1265]: Same here for "#ifndef U_HIDE_DRAFT_API"

File: icu4c/source/i18n/unicode/uspoof.h [hunk 3]
```diff
@@ -1233,11 +1333,48 @@ uspoof_getSkeleton(const USpoofChecker *sc,
                    UChar *dest, int32_t destCapacity,
                    UErrorCode *status);
 
+/**
```
Comment (FrankYFTang) [inline icu4c/source/i18n/unicode/uspoof.h:1336]: Same here for "#ifndef U_HIDE_DRAFT_API"

File: icu4c/source/i18n/unicode/uspoof.h [hunk 4]
```diff
@@ -1273,6 +1410,44 @@ uspoof_getSkeletonUTF8(const USpoofChecker *sc,
                        char *dest, int32_t destCapacity,
                        UErrorCode *status);
 
+/**
```
Comment (FrankYFTang) [inline icu4c/source/i18n/unicode/uspoof.h:1413]: Same here for "#ifndef U_HIDE_DRAFT_API"

File: icu4c/source/i18n/unicode/uspoof.h [hunk 5]
```diff
@@ -1510,11 +1685,40 @@ uspoof_areConfusableUnicodeString(const USpoofChecker *sc,
                                   const icu::UnicodeString &s2,
                                   UErrorCode *status);
 
+/**
```
Comment (FrankYFTang) [inline icu4c/source/i18n/unicode/uspoof.h:1688]: Same here for "#ifndef U_HIDE_DRAFT_API"

File: icu4c/source/i18n/unicode/uspoof.h [hunk 6]
```diff
@@ -1531,7 +1735,7 @@ uspoof_areConfusableUnicodeString(const USpoofChecker *sc,
  *                perform the check.
  * @return        A reference to the destination (skeleton) string.
  *
- * @stable ICU 4.2
```
Comment (FrankYFTang) [inline icu4c/source/i18n/unicode/uspoof.h:1534]: why do you change this line? You should revert this change

File: icu4c/source/i18n/unicode/uspoof.h [hunk 7]
```diff
@@ -1540,6 +1744,34 @@ uspoof_getSkeletonUnicodeString(const USpoofChecker *sc,
                                 icu::UnicodeString &dest,
                                 UErrorCode *status);
 
+/**
```
Comment (FrankYFTang) [inline icu4c/source/i18n/unicode/uspoof.h:1747]: Same here for "#ifndef U_HIDE_DRAFT_API"

File: icu4c/source/i18n/uspoof.cpp [hunk 8]
```diff
@@ -538,6 +539,110 @@ uspoof_areConfusableUnicodeString(const USpoofChecker *sc,
...
+    }
+    if (length1 < -1 || length2 < -1) {
+        *status = U_ILLEGAL_ARGUMENT_ERROR;
+        return 0;
+    }
+    UnicodeString id1Str = UnicodeString::fromUTF8(
+        StringPiece(id1, length1 >= 0 ? length1 : static_cast<int32_t>(uprv_strlen(id1))));
+    UnicodeString id2Str = UnicodeString::fromUTF8(
+        StringPiece(id2, length2 >= 0 ? length2 : static_cast<int32_t>(uprv_strlen(id2))));
+    int32_t results = uspoof_areBidiConfusableUnicodeString(sc, direction, id1Str, id2Str, status);
```
Comment (FrankYFTang) [inline icu4c/source/i18n/uspoof.cpp:575]: you can just return the call instead of storing the value into results then return it.

File: icu4c/source/i18n/uspoof.cpp [hunk 9]
```diff
@@ -697,6 +802,57 @@ uspoof_getSkeleton(const USpoofChecker *sc,
...
+    }
+    if (length < -1 || destCapacity < 0 || (destCapacity == 0 && dest != nullptr)) {
+        *status = U_ILLEGAL_ARGUMENT_ERROR;
+        return 0;
+    }
+
+    UnicodeString idStr((length == -1), id, length); // Aliasing constructor
+    UnicodeString destStr;
+    uspoof_getBidiSkeletonUnicodeString(sc, direction, idStr, destStr, status);
+    destStr.extract(dest, destCapacity, *status);
```
Comment (FrankYFTang) [inline icu4c/source/i18n/uspoof.cpp:821]: I think you can just return the return value of destStr.extract()

File: icu4c/source/i18n/uspoof.cpp [hunk 10]
```diff
@@ -697,6 +802,57 @@ uspoof_getSkeleton(const USpoofChecker *sc,
...
+      return dest;
+    }
+    ubidi_setPara(bidi, id.getBuffer(), id.length(), direction,
+                  /*embeddingLevels*/ nullptr, status);
+    if (U_FAILURE(*status)) {
+        return dest;
+    }
+    UnicodeString reordered;
+    std::int32_t const size = ubidi_getProcessedLength(bidi);
+    ubidi_writeReordered(bidi, reordered.getBuffer(size), size,
```
Comment (FrankYFTang) [inline icu4c/source/i18n/uspoof.cpp:844]: reordered.getBuffer(size) could return nullptr if there are no enough memory, in that case, you should set the 
```
*status = U_MEMORY_ALLOCATION_ERROR;
```
and return;

File: icu4c/source/i18n/uspoof.cpp [hunk 11]
```diff
@@ -697,6 +802,57 @@ uspoof_getSkeleton(const USpoofChecker *sc,
...
+}
+
+
+
+U_I18N_API UnicodeString &U_EXPORT2 uspoof_getBidiSkeletonUnicodeString(const USpoofChecker *sc,
+                                                                        UBiDiDirection direction,
+                                                                        const UnicodeString &id,
+                                                                        UnicodeString &dest,
+                                                                        UErrorCode *status) {
+    UBiDi *bidi = ubidi_open();
```
Comment (FrankYFTang) [inline icu4c/source/i18n/uspoof.cpp:832]: I suggest you call dest.remove(); in the beginning.

File: icu4c/source/i18n/uspoof.cpp [hunk 12]
```diff
@@ -697,6 +802,57 @@ uspoof_getSkeleton(const USpoofChecker *sc,
...
+    ubidi_writeReordered(bidi, reordered.getBuffer(size), size,
+                         UBIDI_KEEP_BASE_COMBINING | UBIDI_DO_MIRRORING, status);
+    reordered.releaseBuffer(size);
+    if (U_FAILURE(*status)) {
+        return dest;
+    }
+
+    ubidi_close(bidi);
+
+    return uspoof_getSkeletonUnicodeString(sc, /*type=*/1729, reordered, dest, status);
```
Comment (FrankYFTang) [inline icu4c/source/i18n/uspoof.cpp:853]: Please define a constant for 1729 to make it clear about the sematic.

File: icu4j/main/tests/core/src/com/ibm/icu/dev/test/text/SpoofCheckerTest.java [hunk 13]
```diff
@@ -470,6 +481,21 @@ void checkSkeleton(SpoofChecker sc, int type, String input, String expected, Str
         assertEquals(testName + " test at line " + lineNumberOfTest + " :  Expected (escaped): " + expected, uExpected, actual);
     }
 
+    // Internal function to run a single skeleton test case.
+    //
+    // Run a single confusable skeleton transformation test case.
+    //
+    void checkBidiSkeleton(SpoofChecker sc, int direction, String input, String expected, String testName) {
+        String uInput = Utility.unescape(input);
```
Comment (FrankYFTang) [inline icu4j/main/tests/core/src/com/ibm/icu/dev/test/text/SpoofCheckerTest.java:489]: Could we fold some of these lines together. Most of them do not need the on stack String. Also, why you need to create t here? junit print these out for you automatically if the assertEquals failed

File: icu4j/main/tests/core/src/com/ibm/icu/dev/test/text/SpoofCheckerTest.java [hunk 14]
```diff
@@ -480,6 +506,18 @@ public void TestAreConfusable() {
...
     }
 
+    @Test
+    public void TestAreBidiConfusable() {
+        SpoofChecker sc = new SpoofChecker.Builder().setChecks(SpoofChecker.CONFUSABLE).build();
+        final String jHyphen2 = "J-2";
+        // The following string has RLMs around the 2–, flipping it; it uses an
+        // EN DASH instead of the HYPHEN-MINUS above.
+        final String j2Dash = "J\u200F2\u2013\u200F";
+        assertEquals("", j2Dash, "J‏2–‏");
```
Comment (FrankYFTang) [inline icu4j/main/tests/core/src/com/ibm/icu/dev/test/text/SpoofCheckerTest.java:516]: put something meaningful in the first parameter of these two lines

File: icu4j/main/classes/core/src/com/ibm/icu/text/SpoofChecker.java [hunk 15]
```diff
@@ -1486,6 +1503,100 @@ public int areConfusable(String s1, String s2) {
...
+            result |= SINGLE_SCRIPT_CONFUSABLE;
+        } else {
+            result |= MIXED_SCRIPT_CONFUSABLE;
+            if (!s1RSS.isEmpty() && !s2RSS.isEmpty()) {
+                result |= WHOLE_SCRIPT_CONFUSABLE;
+            }
+        }
+
+        // Turn off flags that the user doesn't want
+        result &= fChecks;
```
Comment (FrankYFTang) [inline icu4j/main/classes/core/src/com/ibm/icu/text/SpoofChecker.java:1567]: just "return result & fChecks;"

File: icu4c/source/i18n/uspoof.cpp [hunk 16]
```diff
@@ -538,6 +539,109 @@ uspoof_areConfusableUnicodeString(const USpoofChecker *sc,
...
+
+    UnicodeString id1Str((length1 == -1), id1, length1); // Aliasing constructor
+    UnicodeString id2Str((length2 == -1), id2, length2); // Aliasing constructor
+    return uspoof_areBidiConfusableUnicodeString(sc, direction, id1Str, id2Str, status);
+}
+
+U_CAPI uint32_t U_EXPORT2 uspoof_areBidiConfusableUTF8(const USpoofChecker *sc, UBiDiDirection direction,
+                                                      const char *id1, int32_t length1, const char *id2,
+                                                      int32_t length2, UErrorCode *status) {
+    SpoofImpl::validateThis(sc, *status);
```
Comment (FrankYFTang) [inline icu4c/source/i18n/uspoof.cpp:563]: The whole function should be just
```
return uspoof_areBidiConfusableUnicodeString(sc, direction,  
   UnicodeString::fromUTF8(
       StringPiece(id1, length1 >= 0 ? length1 : static_cast<int32_t>(uprv_strlen(id1)))),
   UnicodeString::fromUTF8(
       StringPiece(id2, length2 >= 0 ? length2 : static_cast<int32_t>(uprv_strlen(id2)))),
    status);
```

There are no need for the check in line 563-566 and the check on 567-570 contradict with the API design

File: icu4c/source/i18n/uspoof.cpp [hunk 17]
```diff
@@ -538,6 +539,109 @@ uspoof_areConfusableUnicodeString(const USpoofChecker *sc,
...
+    if (id1RSS.intersects(id2RSS)) {
+        result |= USPOOF_SINGLE_SCRIPT_CONFUSABLE;
+    } else {
+        result |= USPOOF_MIXED_SCRIPT_CONFUSABLE;
+        if (!id1RSS.isEmpty() && !id2RSS.isEmpty()) {
+            result |= USPOOF_WHOLE_SCRIPT_CONFUSABLE;
+        }
+    }
+
+    // Turn off flags that the user doesn't want
```
Comment (FrankYFTang) [inline icu4c/source/i18n/uspoof.cpp:631]: Just
```
return result & This->fChecks;
```

File: icu4c/source/i18n/uspoof.cpp [hunk 18]
```diff
@@ -697,6 +801,64 @@ uspoof_getSkeleton(const USpoofChecker *sc,
...
 
+U_CAPI int32_t U_EXPORT2 uspoof_getBidiSkeleton(const USpoofChecker *sc, UBiDiDirection direction,
+                                                const UChar *id, int32_t length, UChar *dest,
+                                                int32_t destCapacity, UErrorCode *status) {
+
+    SpoofImpl::validateThis(sc, *status);
+    if (U_FAILURE(*status)) {
+        return 0;
+    }
+    if (length < -1 || destCapacity < 0 || (destCapacity == 0 && dest != nullptr)) {
```
Comment (FrankYFTang) [inline icu4c/source/i18n/uspoof.cpp:812]: This check is wrong. Contradict with your API design. Your API doc states
"@param length  .... The length of the input identifier,...., or -1 if the string is zero terminated."
"@param destCapacity .... The destCapacity may be zero... "
You neither need to check the "destCapacity < 0" nor "dest != nullptr" and could just pass to destStr.extract() to let that function to check that. 

The whole function should be implemented as simple as
```
UnicodeString destStr;
uspoof_getBidiSkeletonUnicodeString(sc, direction, UnicodeString((length == -1), id, length), destStr, status);
return destStr.extract(dest, destCapacity, *status);
```

File: icu4c/source/i18n/uspoof.cpp [hunk 19]
```diff
@@ -697,6 +801,64 @@ uspoof_getSkeleton(const USpoofChecker *sc,
...
+      *status = U_ILLEGAL_ARGUMENT_ERROR;
+      return dest;
+    }
+    ubidi_setPara(bidi, id.getBuffer(), id.length(), direction,
+                  /*embeddingLevels*/ nullptr, status);
+    if (U_FAILURE(*status)) {
+        return dest;
+    }
+    UnicodeString reordered;
+    std::int32_t const size = ubidi_getProcessedLength(bidi);
```
Comment (FrankYFTang) [inline icu4c/source/i18n/uspoof.cpp:842]: just int32_t

File: icu4c/source/i18n/uspoof.cpp [hunk 20]
```diff
@@ -697,6 +801,64 @@ uspoof_getSkeleton(const USpoofChecker *sc,
...
+    UChar* const reorderedBuffer = reordered.getBuffer(size);
+    if (reorderedBuffer == nullptr) {
+        *status = U_MEMORY_ALLOCATION_ERROR;
```
Comment (FrankYFTang) [inline icu4c/source/i18n/uspoof.cpp:845]: you leak bidi here. need to close bidi

**Review by FrankYFTang — CHANGES_REQUESTED** [review]
Pleae check your API doc against your checks. Many of your if check inside the C++ implementation are contradicting with your API design and/or redundant/duplicated

File: icu4c/source/i18n/uspoof.cpp [hunk 21]
```diff
@@ -538,6 +539,109 @@ uspoof_areConfusableUnicodeString(const USpoofChecker *sc,
     return result;
 }
 
+U_CAPI uint32_t U_EXPORT2 uspoof_areBidiConfusable(const USpoofChecker *sc, UBiDiDirection direction,
+                                                  const char16_t *id1, int32_t length1,
+                                                  const char16_t *id2, int32_t length2,
+                                                  UErrorCode *status) {
+    SpoofImpl::validateThis(sc, *status);
```
[the remark this answers is not in the document]
Comment (eggrobin) [reply icu4c/source/i18n/uspoof.cpp]: Removed the redundant checks.
However, I do not see how
> The "if (length1 < -1 || length2 < -1) " check is contradict to the API design.

```
 * @param length1 the length of the first identifier, expressed in
 *                16 bit UTF-16 code units, or -1 if the string is
 *                nul terminated.
```
The length of the first identifier cannot be negative, so negative values other than -1 are disallowed.
Likewise, passing a length less than -1 to the UnicodeString constructor [is invalid](https://github.com/unicode-org/icu/blob/667ee72b7c2921d28b935af4e620832054452a09/icu4c/source/common/unistr.cpp#L228-L233) (though UnicodeString [signals errors using the bogus flag rather than error codes](https://github.com/unicode-org/icu/blob/667ee72b7c2921d28b935af4e620832054452a09/icu4c/source/common/unicode/unistr.h#L260-L266)).

Note that the contract is the same as in the old non-bidi version, from which I took those checks.

File: icu4c/source/i18n/uspoof.cpp (hunk 16 above)
Comment (eggrobin) [reply icu4c/source/i18n/uspoof.cpp]: Same as above: removed the redundant `validateThis`, but kept the check for invalid lengths.

File: icu4c/source/i18n/uspoof.cpp (hunk 18 above)
Comment (eggrobin) [reply icu4c/source/i18n/uspoof.cpp]: Removed the `destStr` checks, leaving them to `extract`. Kept the check for invalid `length` (via `isBogus`), as above.

**Review by macchiati — DISMISSED** [review]
I just reviewed the Java and it looks good, someone else should do c/cpp. One javadoc fix

**Review by FrankYFTang — APPROVED** [review]
